truelearn/learning/_base.py

```diff
@@ -0,0 +1,565 @@
+
+        self.draw_proba_type = draw_proba_type
+        self.draw_proba_factor = draw_proba_factor
+        self.draw_proba_static = draw_proba_static
+        self.draw_probability = self.__calculate_draw_proba()
+
+        # create an environment in which training will take place
+        self.__env = trueskill.TrueSkill()
+
+    @final
+    def __calculate_draw_proba(self) -> float:
+        if self.draw_proba_type == "static":
+            # delayed check as this can be potentially replaced by set_params
+            if self.draw_proba_static is None:
+                raise ValueError(
+                    "When draw_proba_type is set to static,"
+                    " the draw_proba_static should not be None."
+                )
+            return self.draw_proba_static * self.draw_proba_factor
+
+        total_engagement_stats = max(
+            1,
+            self.learner_model.number_of_engagements
+            + self.learner_model.number_of_non_engagements,
+        )
+        draw_probability = float(
+            self.learner_model.number_of_engagements / total_engagement_stats
+        )
+
+        # clamp the value between [DEFAULT_DRAW_PROBA_LOW, DEFAULT_DRAW_PROBA_HIGH]
+        draw_probability = max(
+            min(
+                InterestNoveltyKnowledgeBaseClassifier.DEFAULT_DRAW_PROBA_HIGH,
+                draw_probability,
+            ),
+            InterestNoveltyKnowledgeBaseClassifier.DEFAULT_DRAW_PROBA_LOW,
+        )
+
+        # draw_proba_param is a factor if the type is dynamic
+        return draw_probability * self.draw_proba_factor
+
+    @final
+    def __setup_env(self) -> None:
+        """Setup the trueskill environment used in the training process."""
+        self.__calculate_draw_proba()
+        trueskill.setup(
+            mu=0.0,
+            sigma=InterestNoveltyKnowledgeBaseClassifier.DEFAULT_CONTENT_SIGMA,
+            beta=self.beta,
+            tau=self.tau,
+            draw_probability=self.draw_probability,
+            backend="mpmath",
+            env=self.__env,
+        )
+
+    @final
+    def __update_engagement_stats(self, y: bool) -> None:
+        """Update the learner's engagement stats based on the given label.
+
+        Args:
+            y: A bool indicating whether the learner engage in the learning event.
+        """
+        if y:
+            self.learner_model.number_of_engagements += 1
+        else:
+            self.learner_model.number_of_non_engagements += 1
+
+    @final
+    def _gather_trueskill_team(
+        self, kcs: Iterable[AbstractKnowledgeComponent]
+    ) -> Tuple[trueskill.Rating]:
+        """Return a tuple of trueskill Rating \
+        created from the given iterable of knowledge components.
+
+        Args:
+            kcs: An iterable of knowledge components.
+
+        Returns:
+            A tuple of trueskill Rating objects
+            created from the given iterable of knowledge components.
+        """
+        return tuple(
```

Comment:
Expected type 'tuple[Rating]', got 'tuple[Any, ...]' instead

Reply:
It should be `Tuple[trueskill.Rating, ...]`. I don't know why you got `Tuple[Any, ...]`.

Reply:
> It should be `Tuple[trueskill.Rating, ...]`. I don't know why you got `Tuple[Any, ...]`. Let's check it out after I fix this.

After you added the ... the warning has gone

Reply:
Great! When I wrote this code, I may not be very familiar with how to use tuple type hints.